Automattic/woocommerce-payments
# Remove redundant stock reduction attempts

I was not able to reproduce the doubled stock reduction issue mentioned in #3345 in the following cases

1. Saved payment methods as mentioned in https://github.com/Automattic/woocommerce-payments/issues/3345#issuecomment-966104275
2. New payment methods
3. Failing cards
4. Cards that require additional authentication 

All above with UPE enabled/disabled on blocks/classic checkouts. Tried introducing delays in checkout/webhook processing which was also unsuccessful. That said, I noticed a couple of places with redundant stock reduction attempts which this PR proposes to remove.

may fix #3345 

#### Changes proposed in this Pull Request

When the order status changes to processing, `wc_maybe_increase_stock_levels` is invoked, which in turn calls `wc_reduce_stock_levels`. WCPay also call this function manually on the following two occasions
1. When processing a payment 
2. After a successful authentication for payments that require additional requirements

Both above flows also results in a status change, so on the above 2 occasions we attempt to reduce stocks twice. `wc_reduce_stock_levels` should in theory prevent reducing stocks for the same item multiple times, but I am guessing a certain race condition is preventing the above checks. For example if the two calls to `wc_reduce_stock_levels` overlap such that the second stock reduction happens before adding the following meta tag
```$item->add_meta_data( '_reduced_stock', $qty, true );``` it could result in the behavior mentioned in the issue.
Admittedly, I doubt that is really the case here because, as far as I can see, the two calls to `wc_reduce_stock_levels` are always synchronous.

As a precautionary measure, this change removes those extra calls. Also adds a log entry when we are starting to process the order, in order to ease up future debugging attempts to see if we are trying to process the same order twice.


I appreciate any feedback on this or other potential avenues we could explore to reproduce the issue this PR is trying to address









#### Testing instructions





* As a merchant, create a product and enable inventory management
* As a customer, purchase the above product with a saved card and ensure stock is reduced once
* Purchase using a new card and ensure stock is reduced once
* Purchase using a card that require additional authentication (`4000 0027 6000 3184`) and ensure stock is reduced once

-------------------


**Post merge**
